docs/test-plan/stream-subscriber/stream-subscriber-e2e-test-plan.md

```diff
@@ -0,0 +1,912 @@
+# `Stream Subscriber Plugin` Test Plan
+
+## Overview
+
+The `Stream Subscriber Plugin` allows clients to subscribe (issue a request for
+gRPC server streaming) to a Block Node and receive a stream of Blocks based on
+the specified request criteria.
+
+Blocks could be streamed from the Block Node's history and/or from the live
+Block Stream as new Blocks are being published to the Block Node.
+
+## Key Considerations
+
+- **SubscriberServicePlugin**: The plugin accepts gRPC requests from clients
+  and then opens a `BlockStreamSubscriberSession` for each request. Each session
+  runs on its own thread.
+- **BlockStreamSubscriberSession**: Each session handles a single gRPC
+  request from a client on a dedicated thread. The session supplies the Blocks
+  to the client based on the request criteria. Blocks can be supplied from
+  history and/or from the live Block Stream, depending on where the next Block
+  to be sent to the client is located. Going to history and live is
+  interchangeable.
+- **Request Parameters**: A request specifies start and end Block numbers, both
+  inclusive. Both start and end must be whole numbers, as only whole numbers
+  are valid Block numbers. It is possible, however, to also specify -1L
+  (signed long) to both or any of start and end. Note that the gRPC spec uses
+  `uint64` for both values, so `uint64_max` (`0xFFFFFFFFFFFFFFFF`) is the
+  equivalent of -1L signed long. A start of -1L means "start from first
+  available Block", while an end of -1L means "stream indefinitely". When both
+  start and end are specified as -1L, it means that the request is for the next
+  complete live Block, whichever it may be, and then continue to stream
+  indefinitely.
+- **Request Validation**: Each request is validated to ensure that the
+  specified criteria are valid.
+- **Request Fulfillment Precheck**: After a request is validated, a
+  precheck is performed to ensure that the request can be fulfilled. A request
+  that is valid might not be fulfillable at the moment.
+- **Future Requests**: It is possible to subscribe to Blocks that are not yet
+  available in the Block Node. This only applies when the request has specified
+  a start Block. A future request is one where the start Block is greater than
+  the highest Block number currently available to be supplied by the Block Node,
+  be that from history or the live stream. In that case, a check is made if the
+  request is too far into the future. The last permitted start is a value that
+  is calculated by adding a configured maximum future offset to the highest
+  Block number currently available. If the start Block is greater than that
```

Comment:
What resources are held up whiles it waits?

I wonder if it might be simpler to initially start with no future subscribes and mandate that the start block must exist.